// jodit/esm/plugins/ai-assistant/ai-assistant.js
/*!
 * Jodit Editor (https://xdsoft.net/jodit/)
 * Released under MIT see LICENSE.txt in the project root for license information.
 * Copyright (c) 2013-2024 Valeriy Chupurnov. All rights reserved. https://xdsoft.net
 */
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function")
        r = Reflect.decorate(decorators, target, key, desc);
    else
        for (var i = decorators.length - 1; i >= 0; i--)
            if (d = decorators[i])
                r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import watch from "jodit/esm/core/decorators/watch/watch.js";
import { extendLang, pluginSystem } from "jodit/esm/core/global.js";
import { Plugin } from "jodit/esm/core/plugin/plugin.js";
import "./config";
import { UiAiAssistant } from "./ui/ui-ai-assistant";
import * as langs from "./langs";
/**
 * The plugin inserts content generated by AI into the editor.
 */
export class aiAssistant extends Plugin {
    constructor(jodit) {
        super(jodit);
        /** @override */
        this.buttons = [
            {
                name: 'ai-commands',
                group: 'insert'
            },
            {
                name: 'ai-assistant',
                group: 'insert'
            }
        ];
        extendLang(langs);
    }
    /** @override */
    afterInit(jodit) { }
    onGenerateAiAssistantForm(prompt) {
        const { jodit } = this;
        const dialog = jodit.dlg({
            buttons: ['fullsize', 'dialog.close'],
            closeOnClickOverlay: true,
            closeOnEsc: true,
            resizable: false,
            draggable: true,
            minWidth: 460,
            maxWidth: 460
        });
        dialog.bindDestruct(jodit);
        const container = new UiAiAssistant(jodit, {
            onInsertAfter(html) {
                jodit.s.focus();
                jodit.s.setCursorAfter(jodit.s.current());
                jodit.s.insertHTML(html);
                dialog.close();
            },
            onInsert(html) {
                jodit.s.focus();
                jodit.s.insertHTML(html);
                dialog.close();
            }
        });
        container.bindDestruct(dialog);
        dialog.open(container, 'AI Assistant', true, false);
        container.setPrompt(prompt);
    }
    onInvokeAiAssistant(prompt) {
        const { jodit } = this;
        jodit.s.focus();
        const selectedText = jodit.s.html;
        jodit.async
            .promise(async (resolve, reject) => {
            try {
                const result = await jodit.o.aiAssistant
                    .aiAssistantCallback(prompt, selectedText);
                resolve(result);
            }
            catch (error) {
                reject(error);
            }
        })
            .then((htmlFragment) => {
            jodit.e.fire('ai-assistant-response', htmlFragment);
        })
            .catch(error => {
            jodit.message.error(error.message);
            jodit.e.fire('ai-assistant-error', error.message);
        });
    }
    /** @override */
    beforeDestruct(_) { }
}
__decorate([
    watch(':generateAiAssistantForm.ai-assistant')
], aiAssistant.prototype, "onGenerateAiAssistantForm", null);
__decorate([
    watch(':invokeAiAssistant')
], aiAssistant.prototype, "onInvokeAiAssistant", null);
pluginSystem.add('ai-assistant', aiAssistant);
